Add role management procedures

Mats-SX opened the pull request:
changelog [procedures] Add procedures to create and delete roles (`dbms.createRole()` and `dbms.deleteRole()`)

Mats-SX committed:
Add procedure for creation of roles

Mats-SX committed:
Add procedure for deleting roles

Mats-SX committed:
Remove unused class file

Remove unused variables.

Mats-SX committed:
Add per-procedure access mode capabilities

This commit introduces a new attribute to the `Procedure` annotation called `allowed`.
The value of that field is the name of a role in the security system, where any
user that is assigned to that role is given access to execute the procedure in
the procedure's specified access mode, regardless of any other access privileges
owned by any other roles that the user is assigned to.

Mats-SX committed:
Rename `RestrictedAccessMode` to `OverriddenAccessMode`

Mats-SX committed:
Rename method `restrict()` to `overrideWith()`

henriknyman commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthTestBase.java:
```diff
@@ -385,11 +417,33 @@ protected void assertKeyIsMap( ResourceIterator<Map<String, Object>> r, String k
+        @Procedure( name = "test.allowedProcedure1", allowed = "role1", mode = Procedure.Mode.READ )
+        public Stream<AuthProcedures.StringResult> allowedProcedure1()
+        {
+            Result execute = db.execute( "RETURN 'foo' AS foo" );
```
This query does not need READ access to the database (we currently do not have an EXECUTE permission, so you are currently allowed to use Cypher as a calculator with the custom role). If we change the query to use MATCH the test will actually prove something.

henriknyman commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -583,26 +584,60 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+  override def callDbmsProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: String) = {
+    val revertable = transactionalContext.accessMode match {
+      case a: AuthSubject if a.hasRole(allowed) =>
+        Some(transactionalContext.restrictCurrentTransaction(AccessMode.Static.OVERRIDE_SCHEMA))
```
This is not correct. Dbms procedures (currently non-transactional) should not have any different access rights from the user's rights to the database.

henriknyman commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -583,26 +584,60 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
-    callProcedure(name, args, transactionalContext.statement.readOperations().procedureCallRead)
+  override def callReadOnlyProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: String) = {
+    val revertable = transactionalContext.accessMode match {
+      case a: AuthSubject if a.hasRole(allowed) =>
```
I thought we would have a comma separated list of roles to be allowed to give better usability, but I realize that it was not specified on the card.

Mats-SX replied:
Right. Currently it's only possible to specify one role. Does that need to change?

henriknyman commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthTestBase.java:
```diff
@@ -235,6 +249,24 @@ void testFailListRoleUsers( S subject, String roleName, String errMsg )
+        assertFail( subject, "CALL test.allowedProcedure1()", READ_OPS_NOT_ALLOWED );
+        assertFail( subject, "CALL test.allowedProcedure2()", WRITE_OPS_NOT_ALLOWED );
+        assertFail( subject, "CALL test.allowedProcedure3()", SCHEMA_OPS_NOT_ALLOWED );
+//        assertFail( subject, "CALL test.allowedProcedure4()", DBMS_OPS );
```
DBMS procedure access rights are controlled directly on the AuthSubject on a per-procedure basis, and does not map to the static modes for transactions on the database.

henriknyman commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -583,26 +584,60 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+        Some(transactionalContext.restrictCurrentTransaction(AccessMode.Static.OVERRIDE_SCHEMA))
+      case _ => None
+    }
+    callProcedure(name, args, transactionalContext.dbmsOperations.procedureCallDbms(_, _), revertable.foreach(_.close))
```
I realize we need to specify what `allowed` should to for DBMS procedures. Currently the procedures enforce access rights themselves per procedure, so we should probably pass the `allowed` string into procedureCallDbms here so that it can be injected into the procedure context.

Mats-SX replied:
As per our recent discussion, I think we established that the `allowed` flag does not affect `DBMS` procedures in any way.

Mats-SX committed:
Update procedure calls after rebase

fickludd commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthProceduresTestLogic.java:
```diff
@@ -689,6 +694,96 @@ public void shouldAllowAddingAndRemovingUserFromMultipleRoles() throws Exception
+        assertFail( adminSubject, "CALL dbms.security.createRole('&%ss!')",
+                "Role name '&%ss!' contains illegal characters. Use simple ascii characters and numbers." );
+        assertFail( adminSubject, "CALL dbms.security.createRole('åäöø')",
+                "Role name 'åäöø' contains illegal characters. Use simple ascii characters and numbers" );
```
nice to check some utf

fickludd commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthProceduresTestLogic.java:
```diff
@@ -689,6 +694,96 @@ public void shouldAllowAddingAndRemovingUserFromMultipleRoles() throws Exception
+    @Test
+    public void shouldDeleteRole() throws Exception
+    {
+        assertEmpty( adminSubject, format("CALL dbms.security.createRole('%s')", "new_role") );
```
Here one could use `NeoInteractionLevel.getManager` to get a backdoor for creating the role without the full cypher call through the interactionlevel

Mats-SX committed:
Rename `hasRole()` to `allowsProcedureWith()`

Mats-SX committed:
Support naming multiple roles in procedure annotation

The `allowed` attribute is now an array of strings, each mentioning a role name
that gives its users access to the procedure regardless of their normal access privileges.
